src/GovUk.Education.ExploreEducationStatistics.Common/Utils/MultiWriteStream.cs

```diff
@@ -0,0 +1,94 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Utils
+{
+    /// <summary>
+    /// A stream implementation that wraps and writes to multiple target streams.
+    /// <para>
+    /// It should be noted that this stream is only intended for writing,
+    /// and is not intended to be seekable or readable.
+    /// </para>
+    /// </summary>
+    public class MultiWriteStream : Stream
```

Comment:
I think I understand the need for this but just checking. More efficient to start writing the response while creating the zip file at the same time?

Reply:
So the main issue with streams is that they're like a one directional 'pipe' (but not to be confused with `System.IO.PIpe`). You can plug one stream into another stream, but there isn't a nice out of the box way to plug one stream into multiple streams - hence we need something like `MultiWriteStream` that can do exactly that.

For our use case, we're using `MultiWriteStream` to simultaneously:

- Create a temporary zip file so that we can then upload this to blob storage for caching purposes
- Send the zip blob response immediately to the client so they can start downloading it

If we did both of these things serially, this would lead to the client having an unnecessary delay before they can start download the zip.